airbyte-integrations/connectors/source-mixpanel/source_mixpanel/components.py

```diff
@@ -0,0 +1,349 @@
+                    sync_mode=SyncMode.full_refresh, cursor_field=None, stream_state=None
+                ):
+                    empty_parent_slice = True
+                    parent_partition = parent_stream_slice.partition if parent_stream_slice else {}
+
+                    for parent_record in parent_stream.read_records(
+                        sync_mode=SyncMode.full_refresh, cursor_field=None, stream_slice=parent_stream_slice, stream_state=None
+                    ):
+                        # Skip non-records (eg AirbyteLogMessage)
+                        if isinstance(parent_record, AirbyteMessage):
+                            if parent_record.type == Type.RECORD:
+                                parent_record = parent_record.record.data
+                            else:
+                                continue
+                        elif isinstance(parent_record, Record):
+                            parent_record = parent_record.data
+                        try:
+                            partition_value = dpath.util.get(parent_record, parent_field)
+                        except KeyError:
+                            pass
+                        else:
+                            empty_parent_slice = False
+                            yield StreamSlice(
+                                partition={
+                                    partition_field: partition_value,
+                                    "funnel_name": parent_record.get("name"),
+                                    "parent_slice": parent_partition,
+                                },
+                                cursor_slice={},
+                            )
+                    # If the parent slice contains no records,
+                    if empty_parent_slice:
+                        yield from []
+
+
+@dataclass
+class EngagePaginationStrategy(PageIncrement):
+    """
+    Engage stream uses 2 params for pagination:
+    session_id - returned after first request
+    page - incremental page number
+    """
+
+    def next_page_token(self, response, last_records: List[Mapping[str, Any]]) -> Optional[Mapping[str, Any]]:
+        """
+        Determines page and subpage numbers for the `items` stream
+
+        Attributes:
+            response: Contains `boards` and corresponding lists of `items` for each `board`
+            last_records: Parsed `items` from the response
+        """
+        decoded_response = response.json()
+        page_number = decoded_response.get("page")
+        total = decoded_response.get("total")  # exist only on first page
+        if total:
+            self._total = total
+
+        if self._total and page_number is not None and self._total > self.page_size * (page_number + 1):
+            return {"session_id": decoded_response.get("session_id"), "page": page_number + 1}
+        else:
+            self._total = None
+            return None
+
+
+class EngageJsonFileSchemaLoader(JsonFileSchemaLoader):
```

Comment:
why does this need a custom component?

Reply:
@midavadim can you add comments to the code so we we know the intent behind the custom component?